OpenRA.Mods.Common/Traits/Armament.cs

```diff
@@ -251,40 +254,111 @@ protected virtual bool CanFire(Actor self, Target target)
 			return true;
 		}
 
-		// Note: facing is only used by the legacy positioning code
-		// The world coordinate model uses Actor.Orientation
-		public virtual Barrel CheckFire(Actor self, IFacing facing, Target target)
+		// Some advance explanation on Burst behaviors:
+		//
+		// 1) BurstTracking = true + BurstInterruptable = true: the default behavior, namely
+		// 		- firing each burst at the targets' position at the time of firing
+		// 		- using remaining bursts on the next target if target dies before bursts are depleted
+		// 		- does all this by dynamically scheduling and triggering each individual barrel
+		//
+		// 2) BurstTracking = false + BurstInterruptable = false
+		// 		- fires all bursts at the position the target occupied when the barrel firing was scheduled
+		// 		- schedules all barrels at once and uses the schedule delay to emulate regular Burst-/FireDelays.
+		//
+		// 3) BurstTracking = true + BurstInterruptable = false
+		// 		- schedules all barrels at once like above,
+		// 		- but dynamically updates PassiveTarget at the moment of actually firing the barrel to make the weapon 'track' targets
+		//
+		// 4) BurstTracking = false + BurstInterruptable = true
+		// 		- fires all bursts at static position, but cancels unfired bursts if target dies
+		public virtual bool CheckFire(Actor self, Target target)
 		{
 			if (!CanFire(self, target))
-				return null;
+				return false;
 
 			if (ticksSinceLastShot >= Weapon.ReloadDelay)
 				Burst = Weapon.Burst;
 
 			ticksSinceLastShot = 0;
 
-			// If Weapon.Burst == 1, cycle through all LocalOffsets, otherwise use the offset corresponding to current Burst
-			currentBarrel %= barrelCount;
-			var barrel = Weapon.Burst == 1 ? Barrels[currentBarrel] : Barrels[Burst % Barrels.Length];
-			currentBarrel++;
+			if (bulkScheduleBarrels)
+			{
+				var delay = Info.FireDelay;
 
-			FireBarrel(self, facing, target, barrel);
+				// When the weapon uses bulk scheduling, Burst will have been reset by the time the scheduled actions are triggered.
+				// To prevent StartBurstReport from triggering for every burst, FireBarrel needs to know for sure that this is the first burst.
+				// Checking for delay == Info.FireDelay would only work if BurstDelay is > 0, which we can't guarantee in downstream mods.
+				var firstBurst = true;
+				for (var i = Burst; i > 0; i--)
+				{
+					var barrel = Barrels[Burst % Barrels.Length];
+					FireBarrel(self, target, barrel, delay, firstBurst);
```

Comment:
We can remove `firstBurst` by using `FireBarrel(self, target, barrel, delay, i == 0);` instead.